baseplate/observers/metrics_tagged.py

```diff
@@ -0,0 +1,187 @@
+
+    * it tracks the time taken in serving each request.
+    * it batches all metrics generated during a request into as few packets
+      as possible.
+    * it adds tags to the metric if they are in the config tag whitelist
+
+    The batch is accessible to your application during requests as the
+    ``metrics`` attribute on the :py:class:`~baseplate.RequestContext`.
+
+    :param client: The client where metrics will be sent.
+    :param cfg: the parsed application config with the tag whitelist
+
+    """
+
+    def __init__(self, client: metrics.Client, whitelist: List[str]):
+        self.client = client
+        self.whitelist = whitelist
+
+    @classmethod
+    def from_config_and_client(
+        cls, raw_config: config.RawConfig, client: metrics.Client
+    ) -> "TaggedMetricsBaseplateObserver":
+        cfg = config.parse_config(
+            raw_config,
+            {
+                "metrics": {
+                    "whitelist": config.Optional(
+                        config.TupleOf(config.String), default="endpoint,success,error"
+                    ),
+                }
+            },
+        )
+        return cls(client, whitelist=cfg.metrics.whitelist)
+
+    def on_server_span_created(self, context: RequestContext, server_span: Span) -> None:
+        batch = self.client.batch()
+        context.metrics = batch
+        observer = TaggedMetricsServerSpanObserver(batch, server_span, self.whitelist)
+        server_span.register(observer)
+
+
+class TaggedMetricsServerSpanObserver(SpanObserver):
+    def __init__(self, batch: metrics.Batch, server_span: Span, whitelist: list):
+        self.batch = batch
+        self.span = server_span
+        self.base_name = "baseplate.server"
+        self.whitelist = whitelist
+        self.tags = {}  # type: Dict[str, str]
+        self.timer = batch.timer(self.base_name)
+
+    def on_start(self) -> None:
+        self.span.set_tag("endpoint", self.span.name)
+        self.timer.start()
+
+    def on_incr_tag(self, key: str, delta: float) -> None:
+        self.batch.counter(key, self.tags).increment(delta, True)
+
+    def on_set_tag(self, key: str, value: Any) -> None:
+        if key in self.whitelist:
+            if key == "error" and isinstance(value, Errors):
+                self.tags[key] = error_messages[value]
+            else:
+                self.tags[key] = value
+
+    def on_child_span_created(self, span: Span) -> None:
+        observer: SpanObserver
+        if isinstance(span, LocalSpan):
+            observer = TaggedMetricsLocalSpanObserver(self.batch, span, self.whitelist)
+        else:
+            observer = TaggedMetricsClientSpanObserver(self.batch, span, self.whitelist)
+        span.register(observer)
+
+    def on_finish(self, exc_info: Optional[_ExcInfo]) -> None:
+        # elapsed = self.timer.stop_tagged()
+
+        if not exc_info:
+            self.span.set_tag("success", "true")
```

Comment:
i don't think we want to be setting tags on the abstract spans from inside the observers. that'll then cause this observer's `on_set_tag()` to get called which is pretty funky and would lead to ordering issues with the other observers (if we get called after them they'll never see this tag).

Reply:
maybe just add the error stuff directly to the timer's tags?